src/twisted/internet/defer.py

```diff
@@ -53,6 +53,17 @@ class TimeoutError(Exception):
 
 
 
+class NoCurrentResult(BaseException):
+    """
+    This error is raised when attempting to call L{extractCurrentResult} on a
+    L{Deferred} that has not yet called all of its callbacks.
+
+    This error inherits from L{BaseException} rather than L{Exception} to
+    reduce the likelihood of inadvertingly catching this exception as part of
+    an exception handler for other errors.
+    """
```

Comment:
Okay so basically the pattern being anticipated is ...

```
try:
    blah = extractCurrentResult(d)
except Exception as e:
    # Handle the failure result of the Deferred
```

And the idea is to allow `NoCurrentResult` to pass through this because it's not the failure result of `d` but an expectation failure in this example code itself.

Won't the application code want to handle `NoCurrentResult` somehow though?  I'd expect the pattern you'd see much more commonly is ...

```
try:
    blah = extractCurrentResult(d)
except NoCurrentResult:
    # Do ... uh ... something ...
except Exception as e:
    # Handle the failure result of the Deferred
```

Okay, that's not a great example since I can't quite imagine what you'd do in the `NoCurrentResult` handler.  It seems like you'd want to do some kind of translation to a more application-relevant exception.  You don't want `NoCurrentResult` bubbling to the top, killing your program, and dumping a gnarly stack trace at the end user, do you?

Ideally, you really just want this to be a compiler error, in fact... but nevermind.

Also, it seems like `BaseException` suffers from a dilution effect over time.  The more of these exceptions there are (_especially_ if folks actually do let them escape like the first example above does), the more tempting it is to write the first example like ...

```
try:
    blah = extractCurrentResult(d)
except:
    # Handle the failure result of the Deferred
```

which is absolutely awful unless you want to write some code that doesn't get killed by a surprising `BaseException` somewhere. :)

So ... I don't know ... Some sample application code would probably go some way towards revealing whether the second example is legit or not.  If it is, I think `Exception` is an alright base class for this exception.

Reply:
Perhaps code in a loop is a more reasonable example there?

I.e.

```
while True:
    do_dumb_sleep_poll-y_stuff_repeatedly()
    try:
        blah = extractCurrentResult(d)
    except NoCurrentResult:
        continue
    except Exception:
        handle_failure()
    else:
        handle_success()

Reply:
Doesn't that seem like just the kind of use of this API that is wrong and needs to be discouraged?  This isn't how you use Deferreds...

Re-reading this discussion it seems to me like the problem to be solved is "is this an exception from the Deferred I passed in or an exception from the implementation of extractCurrentResult" and I don't think making `NoCurrentResult` a `BaseException` instead of an `Exception` accomplishes this goal.

Instead, putting some more state onto `NoCurrentResult` would make it possible to differentiate.  Attach the Deferred to it (great way to create lots of extra garbage, I guess)?  I'm not sure what else you would do.  This is an example of a perennial problem with Python exceptions - you can't reliably tell where they came from.  This particular API probably isn't going to solve the problem.

Short of some earth-moving idea, I think it makes sense to just switch to subclassing `Exception` and move on.